thunder/dynamo/compiler.py

```diff
@@ -36,27 +310,141 @@ def __init__(self, **thunder_options):
+                    submod_2 = self.submod_2(l_x_, submod_1);  l_x_ = submod_1 = None
+                    return (submod_2,)
+
+                class submod_1(torch.nn.Module):  # Partition for inductor
+                    def forward(self, l_x_: "f32[2]"):
+                        y: "f32[2]" = torch.sinc(l_x_);  l_x_ = None
+                        return y
+
+                class submod_2(torch.nn.Module):  # Partition for thunder
+                    def forward(self, l_x_: "f32[2]", y: "f32[2]"):
+                        matmul: "f32[]" = torch.matmul(l_x_, y);  l_x_ = y = None
+                        return matmul
+        """
+        # Create an `ThunderOperatorSupport` instance which will be used in the callback.
+        # This will determine whether the operation represented by the node is supported by thunder.
+        operator_support = ThunderOperatorSupport(gm)
+
+        # The callback below is called for every node in the graph.
+        # It returns an `int` denoting the parition where the node should be placed.
+        # We want to partition the graph into contiguous regions (with one or more operations)
+        # into thunder supported or unsupported region.
+        # `prev_value` is used to determine if we are still in same region (i.e. supported region or unsupported region).
+        # `partition_cnt` is bumped everytime we change the region i.e. flip from supported to unsupported or from unsupported to supported.
+        # `supported_partitions` is used to track the thunder supported partitions.
+        prev_value = None
+        partition_cnt = 0
+        supported_partitions = set()
+
+        def callback(node) -> int:
+            assert node.op not in (
+                "placeholder",
+                "get_attr",
+                "output",
+            ), f"fx.split_module should have only passed node.op=call_* but received {node.op}"
+            nonlocal prev_value, partition_cnt
+            is_thunder_supported = operator_support.is_node_supported(gm, node)
+            if prev_value == is_thunder_supported:  # We are in the same region.
+                return partition_cnt
+
+            # There is a flip. Either from supported to unsupported or unsupported to supported.
+            prev_value = is_thunder_supported
+            partition_cnt += 1  # Bump the region cnt.
+
+            if is_thunder_supported:
+                supported_partitions.add(partition_cnt)
+            return partition_cnt
+
+        # `fx_split_module` iterates over nodes and determines the partition to place them based on the callback.
+        split_module: torch.fx.GraphModule = fx_split_module(
+            gm, root_m=None, split_callback=callback, keep_original_order=True, keep_original_node_name=True
+        )
+
+        def is_thunder_supported_partition(node: torch.fx.Node) -> bool:
+            return node.name.startswith("submod") and int(node.name.replace("submod_", "")) in supported_partitions
```

Comment:
Instead of "submod" can we use more meaningful names from the start like "thunder_#" and "inductor_#"?

Reply:
To have meaningful names from the start (i.e. from the `split_module` function), we will have to have our own implementation of the same.

However, I am bit a sceptical on doing that as I see there have been few real fixes on the file in the last year or so (not just refactoring and better engineering).
https://github.com/pytorch/pytorch/commits/main/torch/fx/passes/split_module.py

Instead, we can update the graph module returned by `split_module` and rename the submodules so that user will see informative names like `thunder_*` and `inductor_*`. This is what the current code is doing. 

Wanted to know your opinion on the same.

Example of Split subgraph that user would currently see -
```python
class GraphModule(torch.nn.Module):
    def forward(self, l_x_: "f32[2]"):
        # No stacktrace found for following nodes
        thunder_1 = self.thunder_1(l_x_);  l_x_ = None
        return (thunder_1,)
        
    class thunder_1(torch.nn.Module):
        def forward(self, l_x_: "f32[2]"):
             # File: /home/kkalambarkar/lightning-thunder/scratchpad/test_splitter.py:13 in func, code: y = torch.sin(x) + x * (x + 1) + torch.ones(2)
            sin: "f32[2]" = torch.sin(l_x_)
            add: "f32[2]" = l_x_ + 1
            mul: "f32[2]" = l_x_ * add;  add = None
            add_1: "f32[2]" = sin + mul;  sin = mul = None
            ones: "f32[2]" = torch.ones(2)
            y: "f32[2]" = add_1 + ones;  add_1 = ones = None
            
             # File: /home/kkalambarkar/lightning-thunder/scratchpad/test_splitter.py:14 in func, code: return torch.matmul(x, y)
            matmul: "f32[]" = torch.matmul(l_x_, y);  l_x_ = y = None
            return matmul
            
        class _model(torch.nn.Module):
            def forward(self, l_x_: "f32[2]"):
                 # File: /home/kkalambarkar/lightning-thunder/scratchpad/test_splitter.py:13 in func, code: y = torch.sin(x) + x * (x + 1) + torch.ones(2)
                sin: "f32[2]" = torch.sin(l_x_)
                add: "f32[2]" = l_x_ + 1
                mul: "f32[2]" = l_x_ * add;  add = None
                add_1: "f32[2]" = sin + mul;  sin = mul = None
                ones: "f32[2]" = torch.ones(2)
                y: "f32[2]" = add_1 + ones;  add_1 = ones = None
                
                 # File: /home/kkalambarkar/lightning-thunder/scratchpad/test_splitter.py:14 in func, code: return torch.matmul(x, y)
                matmul: "f32[]" = torch.matmul(l_x_, y);  l_x_ = y = None
                return matmul
            
```

Reply:
How much of the file relies on `split_callback(node)` returning an integer? What happens if it starts to return a string that would be used here for the name
https://github.com/pytorch/pytorch/blob/50efbb9f1e7111b4b6d5b8e9a6064ee9783930be/torch/fx/passes/split_module.py#L236

Another option is to patch the Partition class and modify `submod_name` and then restore the original class implementation.
https://github.com/pytorch/pytorch/blob/50efbb9f1e7111b4b6d5b8e9a6064ee9783930be/torch/fx/passes/split_module.py#L18

> Instead, we can update the graph module returned by split_module and rename the submodules so that user will see informative names like thunder_* and inductor_*. This is what the current code is doing.

This is great! No need to come up with workarounds.